cdhweb/events/models.py

```diff
@@ -215,3 +231,177 @@ def ical_event(self):
+    content_object = ParentalKey(
+        "events.Event", on_delete=models.CASCADE, related_name="tagged_items")
+
+
+class Event(Page, ClusterableModel):
+    """Page type for an event, such as a workshop, lecture, or conference."""
+
+    content = StreamField(BodyContentBlock, blank=True)
+    sponsor = models.CharField(max_length=80, null=True, blank=True)
+    start_time = models.DateTimeField()
+    end_time = models.DateTimeField()
+    location = models.ForeignKey(Location, null=True, blank=True,
+                                 on_delete=models.SET_NULL)
+    type = models.ForeignKey(EventType, null=True, on_delete=models.SET_NULL)
+    speakers = models.ManyToManyField(Person, blank=True,
+                                      help_text="Guest lecturer(s) or Workshop leader(s)",
+                                      related_name="events")
+    attendance = models.PositiveIntegerField(null=True, blank=True,
+                                             help_text="Total number of people who attended the event. (Internal only, for reporting purposes.)")
+    join_url = models.URLField(verbose_name="Join URL", null=True, blank=True,
+                               help_text="Join URL for virtual events, e.g. Zoom meetings.")
+    image = models.ForeignKey("wagtailimages.image", null=True, blank=True,
+                              on_delete=models.SET_NULL, related_name="+",
+                              help_text="Image for display on event detail page (optional)")
+    thumbnail = models.ForeignKey("wagtailimages.image", null=True, blank=True,
+                                  on_delete=models.SET_NULL, related_name="+",
+                                  help_text="Image for display on event card (optional)")
+    tags = ClusterTaggableManager(through=EventTag, blank=True)
+    updated = models.DateTimeField(auto_now=True, null=True, editable=False)
+    # TODO attachments (#245)
+
+    # can only be created underneath special link page
+    parent_page_types = ["events.EventsLinkPage"]
+    # no allowed subpages
+    subpage_types = []
+
+    # admin edit configuration
+    content_panels = Page.content_panels + [
+        FieldPanel("type"),
+        FieldRowPanel((FieldPanel("start_time"),
+                       FieldPanel("end_time")), "Details"),
+        FieldPanel("location"),
+        FieldPanel("join_url"),
+        FieldRowPanel((FieldPanel("sponsor"),
+                       FieldPanel("attendance")), "Tracking"),
+        FieldRowPanel((ImageChooserPanel("thumbnail"),
+                       ImageChooserPanel("image")), "Images"),
+        StreamFieldPanel("content"),
+        # FIXME why does this cause recursion errors?
+        # InlinePanel("speakers", label="Speakers")
+    ]
+    promote_panels = Page.promote_panels + [
+        FieldPanel("tags")
+    ]
+
+    # custom manager/queryset logic
+    objects = EventManager()
+
+    context_object_name = "event"
+
+    class Meta:
+        ordering = ("-start_time",)
```

Comment:
would it make sense to add a unique constraint here with a condition so it only applies to events with same month/year? Maybe it should be out of scope, but I think I remember you mentioning this was a bug and I think it might not be that hard see https://docs.djangoproject.com/en/3.1/ref/models/constraints/#uniqueconstraint

Reply:
found a better option, what about using `unique_for_month` and `unique_for_year` on the slug field? 
https://docs.djangoproject.com/en/2.2/ref/models/fields/#unique-for-month

(do we inherit slug from `Page`? wonder if we can re-declare & override it.)

Reply:
oooh, that's really nice. do we need both? i can try redeclaring slug and see what happens; we do inherit it from `Page`

Reply:
nvm I guess we [can't do that](https://docs.djangoproject.com/en/3.1/topics/db/models/#field-name-hiding-is-not-permitted). we could try modifying _attributes_ of the field, although idk what would happen.

Reply:
actually, we might not need to do anything at all - I think wagtail already validates uniquness for slugs at a particular location, and we're forcing all events to exist at the same point in the page tree, so creating a duplicate slug should be impossible

Reply:
confirmed this is the case; wagtail won't let you save it with a duplicate slug even if the dates are different.

Reply:
ok — that seems like it's actually more uniqueness than we actually need (and I wonder if it might lead to weird slugs eventually), but better than allowing url collisions